lib/Backends/OpenCL/kernels.cl

```diff
@@ -183,6 +241,65 @@ size_t getNCHW(ShapeNCHW s, cl_uint32_t n, cl_uint32_t c, cl_uint32_t h,
     name##K(&mem[dest], &mem[lhs], &mem[rhs]);                                 \
   }
 
+/// Macro to define a kernel for data-parallel binary quantized operations. The
+/// body of the kernel is auto-generated by the macro.
+/// \p name the name of the kernel 
+/// \p type the type of the tensor elements and of the return value 
+/// \p body the operation to be performed 
+/// The naming follows the convention of its corresponding implemention in CPU 
+/// baclend.
+#define DEFINE_OPENCL_BINARY_DATA_PARALLEL_KERNEL_QUANTIZED(name, type, body)  \
+  __kernel void name##_i8K(                                                    \
+      __global type *dest, __global type *lhs, __global type *rhs,             \
+      cl_int32_t destOffset, cl_int32_t lhsOffset, cl_int32_t rhsOffset,       \
+      cl_int32_t lhsPre, cl_int32_t lhsPost, cl_int32_t lhsScale,              \
+      cl_int32_t rhsPre, cl_int32_t rhsPost, cl_int32_t rhsScale) {            \
+    size_t i = get_global_id(0);                                               \
+    cl_int32_t LHS =                                                           \
+        scale_i32i8(lhs[i] - lhsOffset, lhsPre, lhsPost, lhsScale, 0);         \
+    cl_int32_t RHS =                                                           \
+        scale_i32i8(rhs[i] - rhsOffset, rhsPre, rhsPost, rhsScale, 0);         \
+    dest[i] = clip((body) + destOffset);                                       \
+  }                                                                            \
+  __kernel void name##_i8W(                                                    \
+      __global void *mem, cl_uint32_t dest, cl_uint32_t lhs, cl_uint32_t rhs,  \
+      cl_int32_t destOffset, QuantizationTransform32To8 lhsScaleParams,        \
+      QuantizationTransform32To8 rhsScaleParams) {                             \
+    name##_i8K(&mem[dest], &mem[lhs], &mem[rhs], destOffset,                   \
+               lhsScaleParams.offset_, rhsScaleParams.offset_,                 \
+               lhsScaleParams.pre_, lhsScaleParams.post_,                      \
+               lhsScaleParams.scale_, rhsScaleParams.pre_,                     \
+               rhsScaleParams.post_, rhsScaleParams.scale_);                   \
+  }
+
+/// Macro to define a mini-kernel for data-parallel multiplicative quantized
+/// operations. The body of the kernel is auto-generated by the macro.
+/// \p name the name of the kernel
+/// \p type the type of the tensor elements
+/// \p body the operation to be performed
+/// The naming follows the convention of its corresponding implemention in CPU 
+/// baclend.
+#define DEFINE_OPENCL_BINARY_DATA_PARALLEL_KERNEL_QUANTIZED_M(name, type,      \
```

Comment:
why do you need to parametrize this by type?

Reply:
it can be used by both ElementMul and ElementDiv, so it's better to parameterize it.

Reply:
If i read this correctly "type" is always int8